### ISSUE TITLE
[SCT_EXTRACT_METRIC] CANNOT USE CUSTOM ATLAS ANYMORE (option -f)

### ISSUE THREAD

**Issue by tanguyduval** [neuropoly/spinalcordtoolbox#1017]
In extract metric, clusters that are used to compute a prioris  are Hard Coded: (Lines 205-212)
`````
    if not len(glob(path_label + 'WMtract*.*')) == 0:
        # MNI-Poly-AMU
        suffix_vertebral_labeling = '*_level.nii.gz'
        ml_clusters = '0:29,30,31'  # 3-class for robust maximum likelihood estimation: WM, GM and CSF
    else:
        # PAM50 and later
        suffix_vertebral_labeling = '*_levels.nii.gz'
        ml_clusters = '0:29,30:35,36'
`````

With an custom atlas with fewer tracts for instance --> fail because cannot find tract 31 to 36.
`````
Remove the temporary folder...[0m
Traceback (most recent call last):
  File "/Users/taduv_admin/codes/spinalcordtoolbox/scripts/sct_extract_metric.py", line 1120, in <module>
    main(fname_data, path_label, method, slices_of_interest, vertebral_levels, fname_output, labels_user, overwrite, fname_normalizing_label, normalization_method, adv_param_user)
  File "/Users/taduv_admin/codes/spinalcordtoolbox/scripts/sct_extract_metric.py", line 331, in main
    indiv_labels_value, indiv_labels_std, indiv_labels_fract_vol = extract_metric(method, data, labels, indiv_labels_ids, ml_clusters, adv_param, normalizing_label, normalization_method)
  File "/Users/taduv_admin/codes/spinalcordtoolbox/scripts/sct_extract_metric.py", line 384, in extract_metric
    clustered_labels, matching_cluster_labels = get_clustered_labels(ml_clusters, labels, list_ids_LOI, combined_labels_id_group, verbose)
  File "/Users/taduv_admin/codes/spinalcordtoolbox/scripts/sct_extract_metric.py", line 1064, in get_clustered_labels
    clustered_labels[i_cluster] = np.sum(labels[clusters_all_labels[i_cluster]])
IndexError: index 31 is out of bounds for axis 1 with size 31

`````

Solutions:
- Use a kmeans clustering to find these clusters
- indicate the clusters in info_label.txt

**Comment by slevyrosetti** [comment]
Indeed, the clusters used as _a priori_ for the MAP estimation should not be hard-coded. They were used to be defined in the `Params` class of sct_extract_metric but with the arrival of the new template, they have been hard-coded to be adapted according to the template version.
In branch `sl_extract_metric` (commit e51c833), I have adapted the code to define the clusters in the info_label.txt file, that thus needs to be adapted. The organization of this file is open to discussion but for the moment it looks like:
~~~
# White matter atlas. Generated on: 01-Sep-2016
# ID, name, file
0, WM left fasciculus gracilis, PAM50_atlas_00.nii.gz
1, WM right fasciculus gracilis, PAM50_atlas_01.nii.gz
2, WM left fasciculus cuneatus, PAM50_atlas_02.nii.gz
3, WM right fasciculus cuneatus, PAM50_atlas_03.nii.gz
4, WM left lateral corticospinal tract, PAM50_atlas_04.nii.gz
5, WM right lateral corticospinal tract, PAM50_atlas_05.nii.gz
6, WM left ventral spinocerebellar tract, PAM50_atlas_06.nii.gz
7, WM right ventral spinocerebellar tract, PAM50_atlas_07.nii.gz
8, WM left rubrospinal tract, PAM50_atlas_08.nii.gz
9, WM right rubrospinal tract, PAM50_atlas_09.nii.gz
10, WM left lateral reticulospinal tract, PAM50_atlas_10.nii.gz
11, WM right lateral reticulospinal tract, PAM50_atlas_11.nii.gz
12, WM left spinal lemniscus (spinothalamic and spinoreticular tracts), PAM50_atlas_12.nii.gz
13, WM right spinal lemniscus (spinothalamic and spinoreticular tracts), PAM50_atlas_13.nii.gz
14, WM left spino-olivary tract, PAM50_atlas_14.nii.gz
15, WM right spino-olivary tract, PAM50_atlas_15.nii.gz
16, WM left ventrolateral reticulospinal tract, PAM50_atlas_16.nii.gz
17, WM right ventrolateral reticulospinal tract, PAM50_atlas_17.nii.gz
18, WM left lateral vestibulospinal tract, PAM50_atlas_18.nii.gz
19, WM right lateral vestibulospinal tract, PAM50_atlas_19.nii.gz
20, WM left ventral reticulospinal tract, PAM50_atlas_20.nii.gz
21, WM right ventral reticulospinal tract, PAM50_atlas_21.nii.gz
22, WM left ventral corticospinal tract, PAM50_atlas_22.nii.gz
23, WM right ventral corticospinal tract, PAM50_atlas_23.nii.gz
24, WM left tectospinal tract, PAM50_atlas_24.nii.gz
25, WM right tectospinal tract, PAM50_atlas_25.nii.gz
26, WM left medial reticulospinal tract, PAM50_atlas_26.nii.gz
27, WM right medial reticulospinal tract, PAM50_atlas_27.nii.gz
28, WM left medial longitudinal fasciculus, PAM50_atlas_28.nii.gz
29, WM right medial longitudinal fasciculus, PAM50_atlas_29.nii.gz
30, GM left dorsal horn, PAM50_atlas_30.nii.gz
31, GM right dorsal horn, PAM50_atlas_31.nii.gz
32, GM left intermediate zone, PAM50_atlas_32.nii.gz
33, GM right intermediate zone, PAM50_atlas_33.nii.gz
34, GM left ventral horn, PAM50_atlas_34.nii.gz
35, GM right ventral horn, PAM50_atlas_35.nii.gz
36, CSF contour, PAM50_atlas_36.nii.gz

# Combined labels
# ID, name, file
50, spinal cord, 0:35
51, white matter, 0:29
52, gray matter, 30:35
52, dorsal columns, 0:3
53, lateral funiculi, 4:13
54, ventral funiculi, 14:29

# Clusters used as a priori for the MAP estimation (for advanced users only)
# Name, file
white matter, 0:29
gray matter, 30:35
CSF, 36
~~~
The k-means clustering could be interesting to define the clusters in a completely "supervised" manner but I am not sure about the robustness of this method and we would need to assess it through simulations, which means a higher amount of work/time for an improvement that does not seem really significant...

**Comment by jcohenadad** [comment]
fixed in PR #1040
